File: generated-code/csharp/codecraft/TcpReadWrite/Codegame/MessageGameModel.cs
namespace TransGenTest.Codegame
{
    /// <summary>
    /// Client or server message
    /// </summary>
    public abstract class MessageGameModel
    {
        /// <summary> Write MessageGameModel to writer </summary>
        public abstract void WriteTo(System.IO.BinaryWriter writer);

        /// <summary> Read MessageGameModel from reader </summary>
        public static MessageGameModel ReadFrom(System.IO.BinaryReader reader)
        {
            switch (reader.ReadInt32())
            {
                case Client.TAG:
                    return Client.ReadFrom(reader);
                case Server.TAG:
                    return Server.ReadFrom(reader);
                default:
                    throw new System.Exception("Unexpected tag value");
            }
        }

        /// <summary>
        /// Client message
        /// </summary>
        public class Client : MessageGameModel
        {
            public const int TAG = 0;
        
            /// <summary>
            /// Message
            /// </summary>
            public Codegame.ClientMessage Message { get; set; }
        
            public Client() { }
        
            public Client(Codegame.ClientMessage message)
            {
                this.Message = message;
            }
        
            /// <summary> Read Client from reader </summary>
            public static new Client ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new Client();
                result.Message = Codegame.ClientMessage.ReadFrom(reader);
                return result;
            }
        
            /// <summary> Write Client to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                Message.WriteTo(writer);
            }
        
            /// <summary> Get string representation of Client </summary>
            public override string ToString() {
                string stringResult = "Client { ";
                stringResult += "Message: ";
                stringResult += Message.ToString();
                stringResult += " }";
                return stringResult;
            }
        }

        /// <summary>
        /// Server message
        /// </summary>
        public class Server : MessageGameModel
        {
            public const int TAG = 1;
        
            /// <summary>
            /// Message
            /// </summary>
            public Codegame.ServerMessage Message { get; set; }
        
            public Server() { }
        
            public Server(Codegame.ServerMessage message)
            {
                this.Message = message;
            }
        
            /// <summary> Read Server from reader </summary>
            public static new Server ReadFrom(System.IO.BinaryReader reader)
            {
                var result = new Server();
                result.Message = Codegame.ServerMessage.ReadFrom(reader);
                return result;
            }
        
            /// <summary> Write Server to writer </summary>
            public override void WriteTo(System.IO.BinaryWriter writer)
            {
                writer.Write(TAG);
                Message.WriteTo(writer);
            }
        
            /// <summary> Get string representation of Server </summary>
            public override string ToString() {
                string stringResult = "Server { ";
                stringResult += "Message: ";
                stringResult += Message.ToString();
                stringResult += " }";
                return stringResult;
            }
        }
    }
}

File: generated-code/fsharp/codecraft/TcpReadWrite/Codegame/MessageGameModel.fs
#nowarn "0058"

namespace TransGenTest.Codegame

open TransGenTest

/// Client message
type MessageGameModelClient = {
    /// Message
    Message: Codegame.ClientMessage;
} with

    /// Write Client to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 0
        this.Message.writeTo writer

    /// Read Client from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Message = Codegame.ClientMessage.readFrom reader;
    }

/// Server message
type MessageGameModelServer = {
    /// Message
    Message: Codegame.ServerMessage;
} with

    /// Write Server to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write 1
        this.Message.writeTo writer

    /// Read Server from reader
    static member readFrom(reader: System.IO.BinaryReader) = {
        Message = Codegame.ServerMessage.readFrom reader;
    }

/// Client or server message
type MessageGameModel =
    /// Client message
    | Client of MessageGameModelClient
    /// Server message
    | Server of MessageGameModelServer
    with

    /// Write MessageGameModel to writer
    member this.writeTo(writer: System.IO.BinaryWriter) =
        match this with
            | Client value -> value.writeTo writer
            | Server value -> value.writeTo writer

    /// Read MessageGameModel from reader
    static member readFrom(reader: System.IO.BinaryReader) =
        match reader.ReadInt32() with
            | 0 -> Client (MessageGameModelClient.readFrom reader)
            | 1 -> Server (MessageGameModelServer.readFrom reader)
            | x -> failwith (sprintf "Unexpected tag %d" x)